presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java

```diff
@@ -0,0 +1,257 @@
+            return IndexSelectionStrategy.none();
+        }
+        String pattern = "^" + Type.THRESHOLD.getCode() + ":\\s*(\\d+)\\s*(%)?$";
+        if (hint.matches(pattern)) {
+            String err = "Invalid 'threshold' hint value '" + hintValue + "' of hint key '" + hintKey + "', regular pattern='" + pattern + "'";
+            Matcher m = Pattern.compile(pattern).matcher(hint);
+            checkState(m.find(), err);
+
+            int integer = Integer.parseInt(m.group(1));
+            String percent = m.group(2);
+
+            if (percent != null) {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT) {
+                    err = "Invalid 'threshold:${maxPercent}%' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithPercent(integer);
+            }
+            else {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS) {
+                    err = "Invalid 'threshold:${maxRows}' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithRows(integer);
+            }
+        }
+        if (hint.matches("^\\[[^\\[\\]]*]$")) {
+            hint = hint.substring(1, hint.length() - 1).trim();
+            if (StringUtils.isBlank(hint)) {
+                return IndexSelectionStrategy.none();
+            }
+            String[] ss = hint.split("[ ;,]+");
+            Set<SchemaTableName> tables = Arrays.stream(ss)
+                    .filter(StringUtils::isNotBlank)
+                    .map(String::trim)
+                    .map(String::toLowerCase)
+                    .map(x -> checkAndAssemblySchemaTableName(currentConnectionSchema, hintKey, x))
+                    .collect(Collectors.toSet());
+            return IndexSelectionStrategy.custom(tables);
+        }
+
+        throw new IllegalArgumentException("Invalid hint value '" + hint + "' of hint key '" + hintKey + "', " + INDEX_HINT_PARSE_ERROR_MESSAGE);
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private static SchemaTableName checkAndAssemblySchemaTableName(Optional<String> currentConnectionSchema,
+            String hintKey,
+            String unit)
+    {
+        String s;
+        if (!unit.contains(".")) { //'schema'.'table'
+            s = currentConnectionSchema.<IllegalArgumentException>orElseThrow(() -> {
+                String str = "Can't obtain the schema of the table[" + unit + "] from current connection for hint '" + hintKey + "'";
```

Comment:
how about inline error message `Can't obtain the schema of`, instead of creating an `str` variable
also, please do not use `str` as variable name, let's try something meaningful